src/LupaClient.php

```diff
@@ -0,0 +1,213 @@
+
+        return $this->httpClient;
+    }
+
+    private function setHttpClient(ClientInterface $httpClient): self
+    {
+        $this->httpClient = $httpClient;
+
+        return $this;
+    }
+
+    private function setJwtToken(?string $jwtToken): self
+    {
+        $this->jwtToken = $jwtToken;
+
+        return $this;
+    }
+
+    private function getJwtToken(): ?string
+    {
+        return $this->jwtToken;
+    }
+
+    public function setEmail(string $email): self
+    {
+        $this->email = $email;
+
+        return $this;
+    }
+
+    public function setPassword(string $password): self
+    {
+        $this->password = $password;
+
+        return $this;
+    }
+
+    /**
+     * @param string $method
+     * @param string $uri
+     * @param bool $requireAuthentication
+     * @param string|null $httpBody
+     * @return array
+     * @throws AuthenticationException
+     * @throws GuzzleException
+     * @throws MissingCredentialsException
+     * @throws TooManyRetriesException
+     */
+    public function send(
+        string $method,
+        string $uri,
+        bool $requireAuthentication = false,
+        ?string $httpBody = null
+    ): array {
+        $request = new Request($method, $uri, self::DEFAULT_HEADERS, $httpBody);
+
+        $attempts = 0;
+        $e = null;
+        do {
+            $attempts++;
+            if ($attempts > self::REQUEST_MAX_RETRIES) {
+                throw new TooManyRetriesException(
+                    'Request failed after ' .
+                        self::REQUEST_MAX_RETRIES .
+                        ' retries',
+                    0,
+                    $e
+                );
+            }
+
+            if ($requireAuthentication) {
+                $jwtToken = $this->getJwtToken();
+                if (!JwtUtils::isJwtTokenValid($jwtToken)) {
+                    $jwtToken = $this->authenticate();
+                }
+
+                $request = $request->withHeader(
+                    'Authorization',
+                    "Bearer $jwtToken"
+                );
+            }
+
+            try {
+                $response = $this->getHttpClient()->send($request);
+
+                return json_decode($response->getBody(), true);
+            } catch (ConnectException $e) {
+                continue;
+            } catch (ClientException $e) {
+                if ($e->getCode() === 401) {
+                    $this->setJwtToken(null);
+                    continue;
+                }
+                throw $e;
+            }
+        } while (true);
+    }
+
+    /**
+     * @return string
+     * @throws AuthenticationException
+     * @throws MissingCredentialsException
+     */
+    public function authenticate(): string
+    {
+        if (!$this->email) {
+            throw new MissingCredentialsException('Email is missing');
+        }
+
+        if (!$this->password) {
+            throw new MissingCredentialsException('Password is missing');
+        }
```

Comment:
If it would be a param in construct - you wont need this part of checking and throwing